labs/00-what-is-tracing/README.md

```diff
@@ -0,0 +1,87 @@
+### What /when/why tracing
+
+Before going further in this tutorial and starting with the labs, let's define tracing: what it is, what it is not, when to use it and how to compare it with other tools.
+
+- - -
+
+#### What is tracing?
+
+`Tracing` consists in recording specific information during a program's or operating system's execution to better understand what is happening on the system. The simplest form of tracing is what we all learn in programming 101: printf!
+
+Every location in the code that we want to trace is called a `tracepoint` and every time a tracepoint is hit, it creates an `event`. The following example shows those concepts
+
+```
+int my_func(void* my_value) {
+  int i, ret;
+
+  printf("Entering my function")                 // <-- tracepoint
+  for (i = 0; i < MAX; i++) {
+    ret = do_something_with(my_value, i);
+    printf("In for loop %d", i);                 // <-- tracepoint
+  }
+  printf("Done: %d", ret);                       // <-- tracepoint
+
+  return ret;
+}
+```
+
+```
+$ ./myprog
+Entering my function                   // <-- event
+In for loop 0                          // <-- event
+In for loop 1                          // <-- event
+In for loop 2                          // <-- event
+Done: 10                               // <-- event
+```
+
+But printf is usually used for quick instantaneous debugging of an application and is removed from production code (right?)
+
+Logging is another form of tracing. It usually associates a timestamp with events to better understand the timing in the application. But we usually log only high level information, as disk space is a limited resource (right?)
+
+The `tracing` we're discussing here is high speed, low overhead tracing. With such tracing, the tracepoints can be present in the code at all time (linux has tons of tracepoints in its code, ready to be hooked to), they have a near-zero overhead when not tracing and a very low one one with a tracer enabled. Tracers can store hundreds of thousands events/second. Try that with printf!!
+
```

Comment:
What about d-trace like aggregators?

Reply:
I think we should differentiate between trace aggregation and trace collection properly. Trace collectors are like `perf trace` LTTng etc while aggregators are like DTrace, `bpftrace` (which uses eBPF)

Reply:
tuxology look at the new version of this paragraph. and see if it's ok wrt that